src/Middleware/RequestBodyParserMiddleware.php

```diff
@@ -0,0 +1,54 @@
+<?php
+
+namespace React\Http\Middleware;
+
+use Psr\Http\Message\ServerRequestInterface;
+use RingCentral\Psr7;
+
+final class RequestBodyParserMiddleware
+{
+    private $keepOriginalBody;
+    private $types = array();
+
+    /**
+     * @param bool $keepOriginalBody Keep the original body after parsing or not
+     */
+    public function __construct($keepOriginalBody = false)
+    {
+        $this->keepOriginalBody = $keepOriginalBody;
+        $this->addType('application/x-www-form-urlencoded', function (ServerRequestInterface $request) {
+            $ret = array();
+            parse_str((string)$request->getBody(), $ret);
+
+            return $request->withParsedBody($ret);
+        });
+    }
+
+    public function addType($type, $callback)
+    {
+        $this->types[$type] = $callback;
+    }
+
+    public function __invoke(ServerRequestInterface $request, $next)
+    {
+        $type = $request->getHeaderLine('Content-Type');
+
+        if (!isset($this->types[$type])) {
+            return $next($request);
+        }
+
+        try {
+            $value = $this->types[$type];
+            /** @var ServerRequestInterface $request */
+            $request = $value($request);
+        } catch (\Exception $e) {
+            return $next($request);
```

Comment:
Should we log somehow the parsing failure?

Reply:
We could allow a `PSR-3` logger as constructor argument and log errors to it, or emit errors as `Server` does

Reply:
It'd be interesting what PHP's behaviour is in case of invalid request bodies. I think we should probably follow that behaviour.